rama-cli/src/cmd/serve/mod.rs

```diff
@@ -0,0 +1,184 @@
+    tracing_subscriber::registry()
+        .with(fmt::layer())
+        .with(
+            EnvFilter::builder()
+                .with_default_directive(LevelFilter::INFO.into())
+                .from_env_lossy(),
+        )
+        .init();
+
+    let maybe_tls_server_config = cfg.secure.then(|| {
+        let tls_key_pem_raw = match std::env::var("RAMA_TLS_KEY") {
+            Ok(raw) => raw,
+            Err(_) => {
+                return ServerConfig {
+                    application_layer_protocol_negotiation: Some(vec![
+                        ApplicationProtocol::HTTP_2,
+                        ApplicationProtocol::HTTP_11,
+                    ]),
+                    ..ServerConfig::new(ServerAuth::SelfSigned(SelfSignedData::default()))
+                };
+            }
+        };
+        let tls_key_pem_raw = std::str::from_utf8(
+            &ENGINE
+                .decode(tls_key_pem_raw)
+                .expect("base64 decode RAMA_TLS_KEY")[..],
+        )
+        .expect("base64-decoded RAMA_TLS_KEY valid utf-8")
+        .try_into()
+        .expect("tls_key_pem_raw => NonEmptyStr (RAMA_TLS_KEY)");
+        let tls_crt_pem_raw = std::env::var("RAMA_TLS_CRT").expect("RAMA_TLS_CRT");
+        let tls_crt_pem_raw = std::str::from_utf8(
+            &ENGINE
+                .decode(tls_crt_pem_raw)
+                .expect("base64 decode RAMA_TLS_CRT")[..],
+        )
+        .expect("base64-decoded RAMA_TLS_CRT valid utf-8")
+        .try_into()
+        .expect("tls_crt_pem_raw => NonEmptyStr (RAMA_TLS_CRT)");
+        ServerConfig {
+            application_layer_protocol_negotiation: Some(vec![
+                ApplicationProtocol::HTTP_2,
+                ApplicationProtocol::HTTP_11,
+            ]),
+            ..ServerConfig::new(ServerAuth::Single(ServerAuthData {
+                private_key: DataEncoding::Pem(tls_key_pem_raw),
+                cert_chain: DataEncoding::Pem(tls_crt_pem_raw),
+                ocsp: None,
+            }))
+        }
+    });
+
+    let maybe_acme_service = std::env::var("RAMA_ACME_DATA")
+        .map(|data| {
+            let mut iter = data.trim().splitn(2, ',');
+            let key = iter.next().expect("acme data key");
+            let value = iter.next().expect("acme data value");
+
+            HijackLayer::new(
+                HttpMatcher::path(format!("/.well-known/acme-challenge/{key}")),
+                AcmeService(value.to_owned()),
+            )
+        })
+        .ok();
+
+    let graceful = rama::graceful::Shutdown::default();
+
+    let tcp_service = ServeServiceBuilder::new()
+        .concurrent(cfg.concurrent)
+        .timeout(Duration::from_secs(cfg.timeout))
+        .maybe_forward(cfg.forward)
+        .maybe_tls_server_config(maybe_tls_server_config)
+        .http_layer(maybe_acme_service)
+        .maybe_content_path(cfg.path)
+        .build(Executor::graceful(graceful.guard()))
+        .expect("build serve service");
```

Comment:
Left all `expect`s for consistency, here output for invalid path is a panic.

Reply:
```suggestion
        .context("build serve service")?;
```

best to return it. You say consistency so I assume you see it somewhere else as well? if so feel free to fix it there. Because a panic renders pretty ugly on a cli. Not critical, but a bit ugly. While we have a nice error catcher in main.rs

Reply:
Sure, in that case I'll also update to handle `AddrInUse` when binding to a port, since that's a common error.
That leaves the setup for TLS and ACME, but I guess that's not expected to fail often.

Reply:
Unless there is a good reason not to we should handle it or propagate it.

Bit out of scope perhaps for your pr but if you have the energy for it, correcting it would be appreciated